dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java

```diff
@@ -316,7 +349,128 @@ public boolean setForceCursorUse(boolean v)
             throw new HpccFileException("Failure sending read ahead transaction", e);
         }
 
-        return buffer;
+        return;
+    }
+
+    public int available() throws IOException
+    {
+        return this.readBufferLen - this.readPos;
+    }
+
+    public void	close() throws IOException
+    {
+        if (this.closed == false)
+        {
+            this.closed = true;
+            this.dos.close();
+            this.dis.close();
+            this.sock.close();
+            this.dos = null;
+            this.dis = null;
+            this.sock = null;
+        }
+    }
+
+    public void	mark(int readlimit)
+    {
+        this.markPos = this.readPos;
+    }
+
+    public boolean	markSupported()
+    {
+        return true;
+    }
+
+    // Returns next byte [0-255] -1 on EOS
+    public int read() throws IOException
+    {
+        boolean onLastByteOfBuffer = this.readPos == this.readBufferLen;
+        if (onLastByteOfBuffer) {
+            try {
+                this.readBlock();
+            } catch (Exception e) {
+                throw new IOException(e.getMessage());
+            }
+        }
+
+        if (this.available() < 1) {
+            return -1;
+        }
+
+        int ret = this.readBuffer[this.readPos];
+        this.readPos++;
+        return ret;
+    }
+
+    // Returns -1 on EOS
+    public int read(byte[] b) throws IOException
+    {
+        return read(b,0,b.length); 
+    }
+
+    // Returns -1 on EOS
+    public int read(byte[] b, int off, int len) throws IOException
+    {
+        boolean needMoreData = (this.readPos + len) >= this.readBufferLen;
+        if (needMoreData) {
+            try {
+                this.readBlock();
+            } catch (Exception e) {
+                throw new IOException(e.getMessage());
+            }
+        }
+
+        if (this.available() < len) {
+            return -1;
+        }
+
+        System.arraycopy(this.readBuffer,this.readPos,b,off,len);
+        this.readPos += len;
+        return len;
+    }
+
+    public void	reset() throws IOException
+    {
+        if (this.markPos < 0) {
+            throw new IOException("Unable to reset to marked position."
+                    + "Either a mark has not been set or the reset length exceeds internal buffer length.");
+        }
+
+        this.readPos = this.markPos;
+    }
+
+    public long	skip(long n) throws IOException
+    {
+        // Have to read the data if we need to skip
+        long remainingBytesToSkip = n;
+        while (remainingBytesToSkip > 0) {
+            if (this.available() == 0 && this.closed) {
+                break;
+            }
+
+            long bytesToSkip = remainingBytesToSkip;
+            if (bytesToSkip > MaxReadSizeKB * 1024) {
```

Comment:
are bytesToSkip/remainingBytesToSkip in MB? If so, let's rename n. Otherwise, should this be if (bytesToSkip * 1024 > MaxReadSizeKB) ?

Reply:
Yep, this is definitely an issue. Corrected in code review changes and added a new test to check reading / writing of large records to exercise this case.